api/genotypes_loader.api.inc

```diff
@@ -91,6 +109,110 @@ function genotypes_loader_helper_add_record_with_mode($record_type, $table, $mod
   }
 }
 
+/**
+ * Places records into a CSV file, and then checks if it is time to bulk load the file via a COPY command
+ *
+ * @param $record_count
+ *   The current count of records that are being loaded in.
+ * @param $record_type
+ *   A human-readable term for your record for error messages.
+ * @param $table
+ *   The chado table to insert into.
+ * @param $insert_values
+ *   An array of [table column] => [value] mappings that is specific to the insert.
+ * @param $file_name
+ *   An optional name for the file to write the values to
+ * @param $final_chunk
+ *   If set to true, then the remaining genotypes have been added to the file and it is time to copy them in.
+ * @return
+ *   TRUE if success (meaning, we reached the end of the function, not neccesarily that everything went swimmingly), FALSE if failed.
+ */
+
+function genotypes_loader_helper_copy_records($record_count, $record_type, $table, $insert_values = array(), $file_name = NULL, $final_chunk = FALSE) {
+
+  // The number of records we want to reach before copying them in all at once
+  $copy_chunk_size = 10000;
+
+  // If we weren't given the name of the file to copy from, then try to find it using the record_type and table name
+  // that we were given
+  if ($file_name === NULL) {
+    $file_stub = '/tmp/genotype_loader.';
+    $record_type = str_replace(' ', '_', $record_type);
+    $record_type = strtolower($record_type);
+    $file_name = $file_stub . $table . '-' . $record_type . '.csv';
+  }
+
+  // If this is the last chunk to be entered into the database, then we don't have a record to add to the file.
+  // Otherwise, open the file to write to since this is needed by fputcsv.
+  // @TODO "Lacey: I'm concerned this might be a point of slowness. Perhaps there is a better
+  //       way to create CSV that doesn't require fputcsv and is more reliable
+  //       then simply imploding the values with comma's in-between."
+  if (!$final_chunk) {
+    // If we don't have a record to load in, something has defintiely gone wrong
+    if ($record_count == 0) {
+      return drush_set_error(dt('There is no genotype call to print to @filename', array('@filename' => $file_name)));
+    }
+    $file_handle = fopen($file_name, 'a') or die ("ERROR: Unable to create $file_name to bulk load records!\n");;
+    ksort($insert_values); // to ensure they're always in the same order.
+    fputcsv($file_handle, $insert_values);
+    fclose($file_handle);
+  }
+
+  if (($record_count%$copy_chunk_size == 0) || ($final_chunk)) {
+    // Write out the COPY command
+    // Columns are listed in alphabetical order because we used ksort() before printing the values to a file.
+    $copy_command = "COPY chado.genotype_call (
+      genotype_id,
+      marker_id,
+      meta_data,
+      project_id,
+      stock_id,
+      variant_id)
+      FROM STDIN WITH CSV";
+
+    genotypes_loader_remote_copy($copy_command, $file_name);
+
+    // Wipe the file clean by reopening it as write-only
```

Comment:
This might be another point of slowness although less of a concern since it only happens once per chunk.